bindings/alicloud/sls/sls.go

```diff
@@ -0,0 +1,126 @@
+package sls
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	sls "github.com/aliyun/aliyun-log-go-sdk"
+	"github.com/aliyun/aliyun-log-go-sdk/producer"
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/kit/config"
+	"github.com/dapr/kit/logger"
+)
+
+type AliCloudSlsLogstorage struct {
+	logger   logger.Logger
+	producer *producer.Producer
+	metadata SlsLogstorageMetadata
+}
+
+type SlsLogstorageMetadata struct {
+	Endpoint        string `json:"endpoint"`
+	AccessKeyID     string `json:"accessKeyID"`
+	AccessKeySecret string `json:"accessKeySecret"`
+}
+
+type Callback struct {
+	s *AliCloudSlsLogstorage
+}
+
+// parse metadata field
+func (s *AliCloudSlsLogstorage) Init(metadata bindings.Metadata) error {
+	m, err := s.parseMeta(metadata)
+	if err != nil {
+		return err
+	}
+	s.metadata = *m
+	producerConfig := producer.GetDefaultProducerConfig()
+	// the config properties in the component yaml file
+	producerConfig.Endpoint = m.Endpoint
+	producerConfig.AccessKeyID = m.AccessKeyID
+	producerConfig.AccessKeySecret = m.AccessKeySecret
+	s.producer = producer.InitProducer(producerConfig)
+
+	s.producer.Start()
+	return nil
+}
+
+func NewAliCloudSlsLogstorage(logger logger.Logger) *AliCloudSlsLogstorage {
+	logger.Debug("initialized Sls log storage binding component")
+	s := &AliCloudSlsLogstorage{
+		logger: logger,
+	}
+	return s
+}
+
+func (s *AliCloudSlsLogstorage) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// verify the metadata property
+	if logProject := req.Metadata["project"]; logProject == "" {
+		return nil, fmt.Errorf("SLS binding error: project property not supplied")
+	}
+	if logstore := req.Metadata["logstore"]; logstore == "" {
+		return nil, fmt.Errorf("SLS binding error: logstore property not supplied")
+	}
+	if topic := req.Metadata["topic"]; topic == "" {
+		return nil, fmt.Errorf("SLS binding error: topic property not supplied")
+	}
+	if source := req.Metadata["source"]; source == "" {
+		return nil, fmt.Errorf("SLS binding error: source property not supplied")
+	}
+
+	log, err := s.parseLog(req)
+	if err != nil {
+		s.logger.Info(err)
+		return nil, err
+	}
+
+	s.logger.Debug(log)
+	callBack := &Callback{}
+	err = s.producer.SendLogWithCallBack(req.Metadata["project"], req.Metadata["logstore"], req.Metadata["topic"], req.Metadata["source"], log, callBack)
+	if err != nil {
+		s.logger.Info(err)
+		return nil, err
+	}
+	return nil, nil
```

Comment:
Is there really nothing we can return to the consumer? Can we return an object like `{"status":"ok"}`? No return at all seems weird

Reply:
the Sls producer api is asynchronous mode, so implemented a callback interface to accept the response of success or faild and record faliure reason to the console log. if return an ok response in the invoke func , may be not ok yet and make the caller misunderstand. 
on the other hand, at most time we may be not so mind whether the log storage is ok…so design to return nil at here.